ivy/src/main/scala/sbt/internal/librarymanagement/IvyUtil.scala

```diff
@@ -1,6 +1,50 @@
 package sbt.internal.librarymanagement
 
+import java.io.IOException
+import java.net.{ SocketException, SocketTimeoutException }
+
+import scala.annotation.tailrec
+import scala.util.{ Failure, Success, Try }
+
 private[sbt] object IvyUtil {
   def separate[A, B](l: Seq[Either[A, B]]): (Seq[A], Seq[B]) =
     (l.flatMap(_.left.toOption), l.flatMap(_.right.toOption))
+
+  @tailrec
+  final def retryWithBackoff[T](
+      f: => T,
+      predicate: Throwable => Boolean,
+      retryLimit: Int,
+      retry: Int = 0
+  ): T = {
+    // Using Try helps in catching only NonFatal exceptions only
+    Try {
+      f
+    } match {
+      case Success(value)                                         => value
+      case Failure(e) if predicate(e) && (retry < retryLimit - 1) =>
+        // max 42 sec backoff
+        val backoff = math.min(math.pow(2d, retry.toDouble).toLong * 1000L, 42000L)
+        Thread.sleep(backoff)
+        retryWithBackoff(f, predicate, retryLimit, retry + 1)
+      case Failure(e) => throw e
+    }
+  }
+
+  object TransientNetworkException {
+    private val _r = """.*HTTP response code: (503|429).*""".r
+
+    @inline private def check(s: String): Boolean = {
+      if (s == null) return false
+
+      _r.pattern.matcher(s).matches()
+    }
+
+    def apply(t: Throwable): Boolean = t match {
+      case _: SocketException | _: SocketTimeoutException => true
+      case e: IOException if check(e.getMessage)          => true
+      case _                                              => false
+    }
```

Comment:
@eed3si9n is there some better way we could identify network errors? This is how gradle does it 
https://github.com/gradle/gradle/blob/master/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/transport/NetworkOperationBackOffAndRetry.java#L58
https://github.com/gradle/gradle/blob/master/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/transport/NetworkingIssueVerifier.java#L27

Reply:
I'm not sure. Whatever you got here seems like a good starting point. We can iterate during patch releases if we need adjustments.